### ISSUE TITLE
Support hashes in base64 notation

### ISSUE THREAD

**Issue by svanderburg** [NixOS/nix#1437]
At the moment, Nix only supports base16 (hexadecimal) and base32 (its own notation) for specifying hash values.

The NPM registry recently upgraded their infrastructure to support SHA512 in base64 notation. Although it's possible to convert hashes to base16 or base32, perhaps it would also be nice to add base64 support so that no conversion is required.

Furthermore, it will also make these sha512 strings in Nix expressions shorter. For example, in base16 notation SHA512 hashes (used for Firefox packages) are quite long.

**Comment by gunpinyo** [comment]
@edolstra I am wondering why base64 is preferred to base64url?
Shouldn't base64url be safer to be in a path?
Anyway, is it possible to add base64url as an alternative notation?

Also, if I understand correctly (please correct me if I am wrong), a custom base32 is used to provide a shorter digest than base16 but doesn't contain slash as in base64; therefore, base64url can even provide a shorter digest yet doesn't contain any special path characters.

**Comment by edolstra** [comment]
@gunpinyo This is the first time I've heard of base64url!

**Comment by gunpinyo** [comment]
@edolstra I am glad that my comment is useful. I hope that you like the idea behind the base64url.

I wish that base64url will become an alternative notation for hashing in nix soon :)

FYI: for a complete reference of base64url please see https://datatracker.ietf.org/doc/html/rfc4648#section-5

TLDR: it is just base64 but digits 62 and 63 change from '+' and '/' to '-' and '_' respectively and padding now become optional.

By the way, I also recommend changing the custom base32 to the standard one in https://datatracker.ietf.org/doc/html/rfc4648#section-6
